python/ray/runtime_env/ARCHITECTURE.md

```diff
@@ -0,0 +1,66 @@
+
+The Raylet's worker pool (src/ray/raylet/worker_pool.cc) handles caching of worker processes and starting new worker processes.  When scheduling a task, its runtime env spec is included in its `TaskSpec`.  The worker pool then compares the hash of the runtime env spec against those of all running workers to determine whether the task can be scheduled on an existing worker process with that runtime env, or whether a new worker process needs to be started.
+
+
+## Creation and deletion
+
+The `RuntimeEnvAgent` exposes gRPC endpoints for `runtime_env` creation and deletion to the Raylet.
+
+The agent manager (src/ray/raylet/agent_manager.cc) runs in the Raylet and manages the connection to the agent, and calls the creation and deletion endpoints.
+
+The worker pool holds a reference to the agent manager, and uses it to send `CreateRuntimeEnvIfNeeded` and `DeleteRuntimeEnvIfPossible` requests to the `RuntimeEnvAgent` as needed for new worker processes or when worker processes are removed. 
+
+## Starting a new worker process
+
+Worker process are started by the Raylet in python/ray/_private/services.py. In this command, python/ray/_private/workers/setup_worker.py is run, which deserializes the `RuntimeEnvContext` object and calls its `exec_worker` method.  The `exec_worker` method sets the appropriate environment variables, modifies the worker process startup command (e.g. prepending `conda activate some_env`) and then calls `execvp` to start the worker process (python/ray/_private/worker.py).
+
+## Caching and garbage collection
+
+For design details on caching and garbage collection, see the [design doc](https://docs.google.com/document/d/1x1JAHg7c0ewcOYwhhclbuW0B0UC7l92WFkF4Su0T-dk/edit#heading=h.j4mqiaz83o96).
+
+The implementation is as follows:
+
+### GCS internal KV garbage collection
+
+This section deals with files that are stored in the head node in the internal KV, such as `working_dir` or `py_modules` packages uploaded by the user.
+
+References are tracked per package (URI).  The reference counting is managed in src/ray/common/runtime_env_manager.cc.  References to these files are incremented when a driver is started that uses that URI, and decremented when a driver exits.  Similarly, references are incremented when a detached actor is started that uses that URI, and decremented when that actor exits.
+
+When a reference count reaches 0, the file is deleted.
+
+### Local node garbage collection
+
+This section deals with files that are stored on disk on all nodes, such as installed `pip` packages or `working_dir` files downloaded from the GCS or from a remote URI.
+
+References for these files are tracked by the runtime env agent process on each node.  Each agent keeps its own reference table in memory.  References are incremented when a runtime env is created, and decremented when the runtime env is deleted.  The files are not deleted until the reference count reaches 0 and the cache size exceeds the maximum cache size.
```

Comment:
max cache size defaults to XXX and can be set using YYY

Reply:
Done